### PR TITLE
[RFR] Use recompose for composing our components and HOC (Follow #107)

### PR DESCRIPTION
**Pull request by djhi**
> Recompose is a React utility belt for function components and higher-order components
[Recompose repository](https://github.com/acdlite/recompose)

### CODE REVIEW

File: .eslintrc [hunk 1]
```diff
@@ -7,6 +7,7 @@
     },
     "extends": "airbnb",
     "rules": {
+        "max-len": "off",
```
Comment (Kmaschta) [inline .eslintrc]: :weary:

**Comment by alexisjanvier** [comment]
👍

File: src/frontend/js/app/withFetchingOnMount.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        }
+
+        render() {
+            const { loading, data, ...props } = this.props;
+
+            if (loading || !data) {
+                return <LoadingComponent />;
+            }
+
+            return factory({
```
Comment (Kmaschta) [inline src/frontend/js/app/withFetchingOnMount.js]: `return factory(props);` ?

Comment (djhi) [inline src/frontend/js/app/withFetchingOnMount.js]: doh

File: src/frontend/js/order/sagas.js [hunk 3]
```diff
@@ -13,26 +14,23 @@ import {
     fetchNewOrder as fetchNewOrderApi,
 } from './api';
 
-export const newOrder = (fetchNewOrder) => function* newOrderSaga({ payload }) {
-    const jwt = yield select(jwtSelector);
-    const {
-        error,
-        order,
-    } = yield call(fetchNewOrder, payload, jwt);
+export const newOrderSaga = fetchNewOrderSaga => function* ({ payload }) {
```
Comment (Kmaschta) [inline src/frontend/js/order/sagas.js]: What about `orderActionsf.failure` action ? It's no longer dispatched with this change

File: src/frontend/js/order/selectors.js [hunk 4]
```diff
@@ -0,0 +1,2 @@
+/* eslint import/prefer-default-export: off */
+export const getOrderById = (state, orderId) => (state.order.list.length ? undefined : state.order.list.find(p => p.id === orderId));
```
Comment (Kmaschta) [inline src/frontend/js/order/selectors.js]: Split in few lines in order to make it more readable

File: src/frontend/js/shoppingcart/reducer.js [hunk 5]
```diff
@@ -14,7 +14,7 @@ export const computeShoppingCartTotal = products =>
     products.reduce((total, p) => total + (p.price * p.quantity), 0);
 
 export default (state = initialState, { type, payload }) => {
-    const products = state.products;
+    const products = [].concat(state.products);
```
Comment (Kmaschta) [inline src/frontend/js/shoppingcart/reducer.js]: What it is for ? Isolation ?
I don't know if `[ ...state.products ]` is preferable, but not very important!

File: src/frontend/js/user/sagas.js [hunk 6]
```diff
@@ -15,28 +16,19 @@ import {
...
-export const signUp = (fetchSignUp, storeLocalUser) => function* signUpSaga({
-    payload: { email, password, previousRoute },
-}) {
-    const { error, user } = yield call(fetchSignUp, email, password);
-    if (error) {
-        yield put(signUpActions.failure(error));
-    } else {
-        yield call(storeLocalUser, user);
-        yield put(signUpActions.success(user));
+export const signUp = (fetchSaga, storeLocalUser) => function* signUpSaga({ payload: { previousRoute, ...payload } }) {
```
Comment (Kmaschta) [inline src/frontend/js/user/sagas.js]: Idem, no `routerActions.failure` ?

**Comment by Kmaschta** [comment]
I did not understand how work recompose yet, but shouldn't we avoid the usage of `withHandlers` ?

**Comment by Kmaschta** [comment]
Oops!

File: src/frontend/js/order/sagas.js (hunk 3 above)
Comment (djhi) [inline src/frontend/js/order/sagas.js]: It is, by the base fetch saga

File: src/frontend/js/order/selectors.js (hunk 4 above)
Comment (djhi) [inline src/frontend/js/order/selectors.js]: forgot to remove this file. The selector is in the reducer file now

File: src/frontend/js/shoppingcart/reducer.js (hunk 5 above)
Comment (djhi) [inline src/frontend/js/shoppingcart/reducer.js]: exactly

File: src/frontend/js/user/sagas.js (hunk 6 above)
Comment (djhi) [inline src/frontend/js/user/sagas.js]: same anwser

**Comment by djhi** [comment]
> I did not understand how work recompose yet, but shouldn't we avoid the usage of withHandlers ?

why ?

**Comment by djhi** [comment]
For info about withHandlers: https://github.com/acdlite/recompose/blob/master/docs/API.md#withhandlers

**Comment by fzaninotto** [comment]
Needs rebase

**Comment by djhi** [comment]
rebased

**Pull request merged**